src/core/search/ast_expr.h

```diff
@@ -4,21 +4,106 @@
 
 #pragma once
 
+#include <algorithm>
+#include <iostream>
+#include <memory>
 #include <ostream>
+#include <vector>
+
+#include "base/logging.h"
 
 namespace dfly {
 
 namespace search {
 
-class AstExpr {};
+// Describes a single node of the filter AST tree.
+class AstNode {
+ public:
+  virtual ~AstNode() = default;
+
+  // Check if this input is matched by the node.
+  virtual bool Check(std::string_view input) const = 0;
+
+  // Debug print node.
+  virtual std::string Debug() const = 0;
+};
+
+using AstNodePtr = std::unique_ptr<AstNode>;
+
+// Do we have to pollute all our parser file with moves?
+struct AstExpr {
```

Comment:
That's the question: auto_ptr style objects would be convenient.

Reply:
I don't think it's that bad, but just my opinion :)

Reply:
you can use shared_ptr if you want to stop moving.

Reply:
Perhaps we don't care much about performance here, but `shared_ptr` is notorious for having some performance artifacts. Like - atomic inc/dec ref for every c'tor/d'tor, as well as doing 2 allocations instead of just ` for `unique_ptr` (unless one uses `make_shared()`, but then it can keep memory unfreed for longer).

I'm not saying we care about this, just that it something to consider :)

Reply:
yes, that's one of the reasons we do not use shared_ptr in Dragonfly :)

In any case, it seems this PR can be left in the firm hands of @chakaz,  so I am removing myself from it 😛 
You boys have fun!